### ISSUE TITLE
Road map for R migration to Node.js

### ISSUE THREAD

**Issue by fxi** [unep-grid/map-x-mgl#757]
## Why a migration is required ? 

_TL;DR;  Long time overdue.  R is not well suited as the back end for our use case. It got better with some async programming options introduced recently but choices have been made and we already use Node.js for the heavy lifting. ._ 

At the beginning of this project in 2014, a static site using Jekyll has been developed to display extractives related maps in a very elegant manner. In 2015, new requirements were needed, such as dynamic statistics and mapping. So, a second mockup,  more functional, has been commissioned. At the time, the excellent Shiny package was the easiest way to build a prototype that would cover many of those aspects on a very short deadline whilst enabling very powerful spacial analysis with R. This solution was not meant to be kept on the long term, as not very scalable: all users interactions blocked the main process no async programming strategy was available at that time. As the project seemed to pivot towards a content management system for geospatial data for a large audience, some strong advice was given to rebuild everything using a more generic web oriented back end, such as Node.js and/or Meteor, that should have scaled better.

Fast forward to 2021, R and the trusty Shiny are still used at the core of this project, mainly to handle registered users interactions ( that's a lot ). Nearly all expensive, blocking operation has been transferred to the Node.js part of MapX.

All what's left are helpers and server-side management, written in R. That's still quite a large part. All those methods ensure reactivity, validation used in users, projects, sources and documents management. Everything is interconnected in an interactive session graph. A lot of optimization  and bugs fixing have been done during all those years. Some parts were simplified, some other made more complex.

As the MapX project took more and more importance over the years, the urge to switch to a more scalable back end is higher now than ever before. 

This migration road map will try to keep a transparent vision of what needs to be done. This will be a very hard and complicated process that will require months of work and be done in parallel to other developments and maintenance activities.

## Reactivity + preliminary jobs

- [x] Integration of full duplex communication between the server and the client (socket.io)
- [x] Implement sample/testing features part 1/3 notification system
- [x] Implement sample/testing features part 2/3 download sources' progress
- [x] Port dictionaries server side for email messaging and other notifications [note: also available in DB : mx_languages]
- [ ] Implement sample/testing features part 3/3 upload sources' progress
- [ ] Serve all built assets directly from nodejs + change paths accordingly client's side
- [ ] Decide if we should keep using reactive programming. It's hard for some dev to visualize mentally a fully reactive tree, especially when she/he did not built it. A more classic approach could be used.
- [ ] Define how how a session object per user/connection should be handled. Ensure that login info, language, roles, projects, etc.. stored in session are fully isolated and functional. 

### Text/translation generation
Besides warnings and other notifications, some texts are built / translated in the app. As all translations are pushed in the client app when a language is chosen, this could be something to port to the client. It concerns all parts where text is returned, especially JSON templates. 

- [ ] Port the translation system from client to server, as some UI elements have to be built server side.
- [ ] Test @fxi/el package server side for UI building 

# Main parts to migrate

- [x] Geoserver ( interact with geoserver via a client. high priority. If too much friction, this could compromise the whole migration, as our current client geosapi, works well and no alternatives have been found in nodes, besides geoserver-node-client, which is still in alpha )
- [ ] JSON templating
- [ ] Login process
- [ ] Projects
- [ ] Views
- [ ] Sources
- [ ] Language
- [ ] Map

# Files/modules removal/migration

### root
- [ ] server.R
- [ ] global.R
- [ ] run.R

### helpers

- [ ] api_fetch.R
- [ ] binding_epsgio.R
- [ ] binding_jed.R
- [ ] binding_mgl.R
- [ ] binding_mx.R
- [ ] binding_wms_client.R
- [ ] db.R
- [x] db_log.R
- [ ] db_pool.R
- [ ] db_projects.R
- [ ] db_views.R
- [ ] email_sender.R
- [x] geoserver.R
- [ ] login_utils.R
- [ ] misc.R
- [ ] query_parser.R
- [ ] schema_source_meta.R
- [ ] schema_view_story.R
- [ ] schema_view_style.R
- [ ] story_auto_start.R
- [x] wms.R

### server
- [ ] controls.R
- [ ] input_register.R
- [ ] ip.R
- [ ] language.R
- [ ] login.R
- [ ] main.R
- [ ] maintenance.R
- [ ] map.R
- [ ] project.R
- [ ] react_ip_data.R
- [ ] react_login_key_status.R
- [ ] react_map.R
- [ ] react_objects.R
- [ ] react_project_members.R
- [ ] react_roles.R
- [ ] react_source_list.R
- [ ] react_source_summary.R
- [ ] react_view_list.R
- [ ] root_mode.R
- [x] share.R
- [ ] share_to_project.R
- [x] source_download.R
- [x] tools_app_config.R
- [x] tools_db_connect.R
- [ ] tools_project_config.R
- [ ] tools_project_delete.R
- [ ] tools_project_external_views.R
- [ ] tools_project_invite.R
- [ ] tools_project_join.R
- [ ] tools_project_main.R
- [x] tools_project_new.R
- [ ] tools_project_roles.R
- [ ] tools_project_sources_views.R
- [ ] tools_project_views_state.R
- [x] tools_query_maker.R
- [ ] tools_source_edit_metadata.R
- [ ] tools_source_edit_table.R
- [ ] tools_source_manage.R
- [ ] tools_source_new.R
- [ ] tools_source_overlap.R
- [x] tools_source_validate_geom.R
- [ ] tools_view_new.R
- [ ] user.R
- [ ] view_edit.R
- [ ] view_edit_custom.R
- [ ] view_edit_dashboard.R
- [ ] view_edit_story_map.R
- [ ] view_edit_style.R
- [ ] view_update_client.R
- [x] wms_creator.R

### scripts
- [x] updateGeoserverSource.R
- [x] updateGeoserverView.R

**Comment by fxi** [comment]
State of the migration, august 2022. @PierreLacroix 

I wrote a small script to extract code stat for the R language, over time.

- Red line = Server/Logic code. This should go down over time. The more features are transferred to node.js, the less logic remains in the R/Shiny side.
- Black line = Helpers/Library code. This will remain stable until the end. If a block of logic is removed and a set of helpers is only used there, progressive steep drop will be seen.

Comments : 
- Since the implementation of the API in Node.js, slow improvement, even if the app complexity is growing.  
- The removal of GeoServer helpers was a big step forward, but  the goal is still far away. 
- Some new feature requires minimal R bindings, easy to migrate later. This is what is visible recently.
- If some features rely heavily on the APP logic are requested, the R side will continue to grow. 
- The development of the authentication / validation in node, via websocket will ease the transition. 
- Still a slow process, unless a large amount of time is dedicated exclusively to this task.
- The use of docker swarm, code and database optimization, and migration prioritization based on computing cost or blocking issue helped tremendously the overall performance gain.  


<img width="1516" alt="image" src="https://user-images.githubusercontent.com/1196833/183043893-3740788a-cda0-4eeb-b7e7-ae18b730c8f7.png">
